### DISCUSSION TITLE
Windows 웹브라우저에서 발생하는 글리치 사운드 제거 (euphony.js)

### DISCUSSION THREAD

**Discussion by kuro11pow2** [euphony-io/euphony#90] [General]
> 정보 공유해요

[euphony.js](https://github.com/euphony-io/euphony.js)

문제: Windows 환경의 웹브라우저에서 발생하는 글리치 사운드 = 삐리리리 소리 (원래 안 들림)
재현: [소개 사이트](https://dev.jbear.co/euphony/)에서 demo > broadcast 로 확인하실 수 있습니다.

맥북에서는 발생하지 않는 현상이라고 합니다. 왜 윈도우즈에서만 생길까요?
테스트 결과 출력 장치의 종류와 무관하게 발생하는 것으로 보입니다.
(내장 스피커, 유선 스피커, 블루투스 스피커, 블루투스 이어폰에서 모두 발생)

https://github.com/euphony-io/euphony.js/blob/master/euphony.js#L23-L30
```js
        this.BUFFERSIZE = 2048;
        this.PI = 3.141592653589793;
        this.PI2 = this.PI * 2;
        this.SAMPLERATE = 44100;
        this.SPAN = 86;
        this.BASE_FREQUENCY = 18000;
        this.CHANNEL = 1;
        this.setModulation('CPFSK');
```
```js
    euphony.prototype = {
        initBuffers: function () {
            const T = this;


            T.scriptProcessor = null;
            T.outBuffer = new Array();
            T.playBuffer = new Array();
            T.playBufferIdx = 0;


            T.startPointBuffer = T.crossfadeStaticBuffer(T.makeStaticFrequency(T.BASE_FREQUENCY - T.SPAN));
            T.zeroBuffer = T.makeZeroSource();
            /*
              CHANNEL 1 : 0 ~ 16
              CHANNEL 2 : 17 ~ 32
            */
            for (let i = 0; i < 32; i++) {
                T.outBuffer[i] = T.crossfadeStaticBuffer(T.makeStaticFrequency(T.BASE_FREQUENCY + i * T.SPAN));
            }


            T.progressIdx = 0;
            T.cp_index = 0;
            T.cp_last_theta = 0;
        },
```

주파수가 BASE_FREQUENCY  + n * SPAN 꼴로 설정되는 것으로 보입니다.
18000이면 확실히 거의 들리지 않아야 할 대역인데 너무 잘 들리네요

**Comment by kuro11pow2** [comment]
기한이 갑자기 절반이나 줄었네요 ㅠ  js에서도 Charset 관련 기여를 하는 게 맞을 것 같습니다 ㅠㅠ 
그래도 찾아본 게 아쉬워서 공부 자료라도 공유할게요

[oscillator](https://developer.mozilla.org/en-US/docs/Web/API/OscillatorNode)
[web audio api](https://developer.mozilla.org/ko/docs/Web/API/Web_Audio_API/Using_Web_Audio_API)

**Reply by aiclaudev** [reply]
정말 아쉽습니다ㅠㅠ 자료 감사합니다!!

**Comment by YoungSeokHong** [comment]
오 저도 이슈를 하나 더 해결해보고 싶었는데, 한 번 같이 해결해 봅시다!
저도 참여해서 열심히 조사해서 정보공유하겠습니다!
기한이 갑자기 줄어서 아쉽긴 하네요 ㅠㅠ 그래도 마지막까지 한 번 화이팅해봅시다.🚀

**Reply by kuro11pow2** [reply]
감사합니다 아래 샘플 코드도 올려두었습니다! 같이 정보 공유해봐요 😄

**Reply by aiclaudev** [reply]
😃👍

**Comment by kuro11pow2** [comment]
주파수 출력 샘플 코드 공유합니다
start 버튼을 누르고 마우스를 좌우로 움직이면 0~20000hz 사이의 출력을 확인할 수 있습니다.
mute unmute 가능합니다.

오실레이터로 만드는 주파수는 정상적으로 출력되는 것 같습니다.  euphony.js에서도 아래 코드와 마찬가지로 AudioContext를 사용하는 것으로 보이는데 어떤 부분에서 문제가 발생하는지는 더 살펴봐야할 것 같네요 ㅠ 
euphony.js에서는 euphonyWorkletNode나 buffer를 사용하는 부분이 다르긴 한데 여기서 생기는 문제일까요..

test.html
```html
<!DOCTYPE html>
<html>

<head>
    <meta charset='utf-8'>
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title> demo </title>
    <script src="test.js"></script>
</head>

<body>
    <center>
        <button class="play">시작</button>
        <button class="mute" data-muted="false">mute</button>
        <div class="freq">0</div>
    </center>
</body>

</html>
```
test.js
```js
window.onload = function () {
    document.querySelector('.play')
        .addEventListener('click', init);
}

let isAppInit = false;
const init = function () {
    if (isAppInit) {
        return;
    }
    // create web audio api context
    const AudioContext = window.AudioContext || window.webkitAudioContext;
    const audioCtx = new AudioContext();

    // create Oscillator and gain node
    const oscillator = audioCtx.createOscillator();
    const gainNode = audioCtx.createGain();

    // connect oscillator to gain node to speakers
    oscillator.connect(gainNode);
    gainNode.connect(audioCtx.destination);

    // create initial theremin frequency and volume values
    const WIDTH = window.innerWidth;
    const HEIGHT = window.innerHeight;

    const maxFreq = 20000;
    const initialVol = 0.015;

    // set options for the oscillator
    oscillator.detune.value = 100; // value in cents
    oscillator.start(0);

    oscillator.onended = function () {
        console.log('Your tone has now stopped playing!');
    };

    gainNode.gain.value = initialVol;
    gainNode.gain.minValue = initialVol;
    gainNode.gain.maxValue = initialVol;

    const mute = document.querySelector('.mute');

    mute.onclick = function () {
        if (mute.getAttribute('data-muted') === 'false') {
            gainNode.disconnect(audioCtx.destination);
            mute.setAttribute('data-muted', 'true');
            mute.innerHTML = "Unmute";
        } else {
            gainNode.connect(audioCtx.destination);
            mute.setAttribute('data-muted', 'false');
            mute.innerHTML = "Mute";
        };
    }

    document.onmousemove = updatePage;

    const freq = document.querySelector('.freq');

    function updatePage(e) {
        KeyFlag = false;

        CurX = e.pageX;
        CurY = e.pageY;

        oscillator.frequency.value = (CurX / WIDTH) * maxFreq;
        freq.textContent = Math.round(oscillator.frequency.value);
    }

    isAppInit = true;
}
```

**Reply by aiclaudev** [reply]
코드 해석해봐야겠네요!! 수고하셨습니다 ㅎㅎ

**Comment by designe** [comment]
@kuro11pow2 
거의 다 오신 것 같은데요!? ㅎㅎ

**Comment by designe** [comment]
참고로 주로 사용되는 Samplerate는 44100과 48000이 있습니다 :)
맥에서는 44100이 메인 Samplerate인데, Windows는 어떤지 궁금하네요.

혹시 샘플레이트에 대해서 한번 정리해서 discussion 올려주실 분 있나요?

**Reply by ycs1m1yk** [reply]
Windows 10 노트북 2대에서 확인해보니 48000이 기본 Samplerate 인것 같습니다!

<img width="450" alt="windows10-default-sr" src="https://user-images.githubusercontent.com/27720475/136069902-504d9f30-e4cc-4468-8b52-fa39ae7189df.png">

**Reply by kuro11pow2** [reply]
스피커의 samplerate를 44100Hz으로 변경하니 소리가 안 들리네요!
그런데 조금 이상한 점이

```
스피커 samplerate = 44100Hz 
this.SAMPLERATE = 44100
```
여기서는 소리가 안 들리는데

```
스피커 samplerate = 48000Hz 
this.SAMPLERATE = 48000
```
여기서는 소리가 들립니다

**Reply by ycs1m1yk** [reply]
페이지 새로고침하고 다시 시도해도
소리가 발생하나요?

**Reply by kuro11pow2** [reply]
아하 알아냈습니다!!

아래처럼 사용자 환경의 samplerate를 가져오면 됩니다!

```js
    function euphony() {
        ...
        this.context = new (window.AudioContext || window.webkitAudioContext)();
        this.SAMPLERATE = this.context.sampleRate;
        ...
    };
```

추가적으로 알아낸 부분이 몇 가지 있습니다.

1. 스피커의 samplerate를 수정한 경우 브라우저를 새로고침해야 적용됩니다.
2. 사용자가 오디오 인터페이스를 사용중이면 방법이 없습니다.

제 경우에는 가상 오디오 인터페이스를 사용중이었는데 이 친구가 자기 마음대로 샘플레이트를 바꿔버리고 있더라구요
![image](https://user-images.githubusercontent.com/47289893/136121008-923ea4ea-be6a-4048-b5ee-f662eb1fd438.png)

**Reply by kuro11pow2** [reply]
> 페이지 새로고침하고 다시 시도해도 소리가 발생하나요?

제 경우에는 가상 오디오 인터페이스가 문제였던 것 같습니다 
이 친구가 매번 말썽이네요 ㅋㅋ큐ㅠㅠ

**Reply by designe** [reply]
+1

**Reply by aiclaudev** [reply]
엄청나네요.. 대단하십니다👍👍

**Comment by ycs1m1yk** [comment]
스피커 속성에서 Sample rate(샘플 속도)를 44100Hz, 48000Hz로 설정했을때
44100Hz에서는 글리치 사운드가 나지 않고, 48000Hz에서는 발생하는 현상을 확인했습니다.

### 스피커 Sample rate 설정법

1. 작업표시줄 우측의 스피커 아이콘 **우클릭**

<img width="450" alt="speaker-icon" src="https://user-images.githubusercontent.com/27720475/136070770-04131627-3d0a-447c-8d6b-bb5fb9813a21.png">

2. **소리(s)** 메뉴 클릭

<img width="450" alt="sound-menu" src="https://user-images.githubusercontent.com/27720475/136070785-35830b79-b50c-416a-b3e5-cf84f75eaca3.png">

3. 사용중인 장치를 선택하고 우측하단 **속성** 클릭

<img width="450" alt="properties-menu" src="https://user-images.githubusercontent.com/27720475/136071296-1d9a98ea-b1d7-4299-9695-b036d65d3a50.png">

<img width="450" alt="properties" src="https://user-images.githubusercontent.com/27720475/136070868-26bd8954-c580-4b5a-aad6-0ac1ce565ecb.png">

4.  **고급** 탭으로 이동

<img width="450" alt="advanced-tab" src="https://user-images.githubusercontent.com/27720475/136070985-dfc6eaa0-0b2d-43c0-9b1a-3ce5df4e965c.png">

5. **기본 형식**에서 Sample rate 변경

<img width="450" alt="change-sr" src="https://user-images.githubusercontent.com/27720475/136071166-0a5a9396-7a56-47d0-9890-b52e5847502e.png">

**Reply by ycs1m1yk** [reply]
`euphony.js`에서 SAMPLERATE = 48000 으로 하드코딩하고 테스트해보니

이번에는 반대로 48000Hz에서는 글리치 사운드가 나지 않고, 44100Hz에서는 발생하는 현상을 확인했습니다!

<img width="450" alt="48000" src="https://user-images.githubusercontent.com/27720475/136078178-8a93ec37-a99d-4d28-8db4-fa0bd7cc711e.png">

**Reply by aiclaudev** [reply]
와우 정리 감사합니다!!

**Comment by judemin** [comment]
매일매일 새로운게 업데이트 되네요...다들 정말 대단하십니다 👍

**Reply by designe** [reply]
:)

**Comment by aiclaudev** [comment]
코멘트들만 쭉 읽어봐도 큰 공부가 될 것 같아요. 감사합니다!! :)

**Comment by YoungSeokHong** [comment]
어제 알아보다가 오늘 아침부터 수업이 있어서 일찍 잤는데 일어나 보니 하루도 안되서 진도가 엄청 나가있네요 ㅋㅋㅋㅋㅋㅋㅋㅋ
정말 너무 대단하십니다ㅠㅠ 저도 코멘트들 읽어보면서 먼저 가신 길 따라가봐야겠습니다!

**Comment by YoungSeokHong** [comment]
저는 맥북과 윈도우로 모든 인터넷 브라우저에서 "123123"을 갖고 http://localhost:8080/demo/ 와 [Euphony Listener](https://github.com/euphony-io/euphony-listener.git) 으로 확인을 해봤습니다.


<table>
 <tr>
    <td>OS</td>
    <td>브라우저</td>
    <td>소리</td>
    <td>작동</td>
  </tr>
  <tr>
    <td rowspan="3">Mac</td>
    <td>사파리</td>
    <td>아무소리 안남</td>
    <td>시간도 오래걸리고 이상하게 출력됨</td>
  </tr>
  <tr>
    <td>구글 크롬</td>
    <td>삐리리- 소리가 남 (잡음 없이 깔끔)</td>
    <td>정상적으로 작동</td>
  </tr>
  <tr>
    <td>네이버 웨일</td>
    <td>삐리리- 소리가 남 (잡음 없이 깔끔)</td>
    <td>정상적으로 작동</td>
  </tr>

  <tr>
    <td rowspan="3">Window</td>
    <td>엣지</td>
    <td>삐리리- 소리가 남(지지직 잡음 포함)</td>
    <td>정상적으로 작동</td>
  </tr>
  <tr>
    <td>구글 크롬</td>
    <td>삐리리- 소리가 남(지지직 잡음 포함)</td>
    <td>정상적으로 작동</td>
  </tr>
  <tr>
    <td>네이버 웨일</td>
    <td>삐리리- 소리가 남(지지직 잡음 포함)</td>
    <td>정상적으로 작동</td>
  </tr>
</table>

맥북의 사파리를 제외하고는 다 정상적으로 작동 했지만 삐리리- 소리는 계속해서 들렸고 윈도우의 경우는 추가적으로 지지직하는 잡음도 들렸습니다. 무시할만한 잡음이긴 했지만 맥북하고 바로 비교했을 때는 잡음이 들린다는 것을 확실히 알 수 있을 정도였습니다.
사파리의 경우에는 "123123"을 input으로 줬는데, 시간도 3분 이상 걸렸고 아래와 같은 결과가 나왔습니다. 사파리의 결과를 보았을 때는 SAMPLERATE이 아니라 인코딩 방식의 차이 때문인것 같은데 더 찾아봐야겠습니다!

<img src="https://user-images.githubusercontent.com/80624082/136235792-ee8ecd7b-753b-4f0d-89c7-3395a4dc181d.jpg" width="400">

@ycs1m1yk @kuro11pow2 준규님, 혜성님 말씀대로 euphony.js에서 SAMPLERATE을 변경한 뒤에 시도 해봤을 때는 잡음이 확실히 잡히고 소리도 더 커졌습니다!

두 분 정말 고생하셨습니다. 감사합니다!

**Reply by ycs1m1yk** [reply]
다양한 브라우저 환경에서 테스트 감사합니다!
고생하셨네요ㅎㅎ

**Reply by kuro11pow2** [reply]
@YoungSeokHong 
와 정말 다양한 환경에서 테스트해주셨네요 고생하셨습니다!!

**Reply by aiclaudev** [reply]
와우 수고많으셨습니다~!

**Reply by dlwlstks96** [reply]
고생 많으셨습니다!!!🎉🎉

**Comment by seohy02** [comment]
저도 계속 방법 찾아볼게요 !!! 도움이 많이 되었습니다!😀😀😀👍👍👍

**Reply by kuro11pow2** [reply]
감사합니다ㅎㅎ 😄

**Comment by designe** [comment]
매우매우 잘하고 계십니당 ^^ 정리하셔서 PR로 기여해주시지요

**Reply by kuro11pow2** [reply]
넵 정리해서 PR 올리겠습니다! 👍
